src/Tools/dotnet-monitor/CollectionRules/Actions/CollectDumpAction.cs

```diff
@@ -16,67 +16,127 @@
         private readonly IServiceProvider _serviceProvider;
 
-        public CollectDumpAction(IServiceProvider serviceProvider)
+        public CollectDumpActionFactory(IServiceProvider serviceProvider)
         {
             _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
-            _dumpService = serviceProvider.GetRequiredService<IDumpService>();
         }
 
-        public async Task<CollectionRuleActionResult> ExecuteAsync(CollectDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        public ICollectionRuleAction Create(IEndpointInfo endpointInfo, CollectDumpOptions options)
         {
-            if (options == null)
+            if (null == options)
             {
                 throw new ArgumentNullException(nameof(options));
             }
 
-            if (endpointInfo == null)
-            {
-                throw new ArgumentNullException(nameof(endpointInfo));
-            }
+            ValidationContext context = new(options, _serviceProvider, items: null);
+            Validator.ValidateObject(options, context, validateAllProperties: true);
 
-            DumpType dumpType = options.Type.GetValueOrDefault(CollectDumpOptionsDefaults.Type);
-            string egressProvider = options.Egress;
+            return new CollectDumpAction(_serviceProvider, endpointInfo, options);
+        }
 
-            string dumpFileName = Utils.GenerateDumpFileName();
+        internal sealed class CollectDumpAction :
+            ICollectionRuleAction,
+            IAsyncDisposable
+        {
+            private readonly CancellationTokenSource _disposalTokenSource = new();
+            private readonly IDumpService _dumpService;
+            private readonly IEndpointInfo _endpointInfo;
+            private readonly CollectDumpOptions _options;
+            private readonly IServiceProvider _serviceProvider;
 
-            string dumpFilePath = string.Empty;
+            private Task<CollectionRuleActionResult> _completionTask;
 
-            ValidationContext context = new(options, _serviceProvider, items: null);
-            Validator.ValidateObject(options, context, validateAllProperties: true);
+            public CollectDumpAction(IServiceProvider serviceProvider, IEndpointInfo endpointInfo, CollectDumpOptions options)
+            {
+                _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+                _options = options ?? throw new ArgumentNullException(nameof(options));
+                _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
+                _dumpService = serviceProvider.GetRequiredService<IDumpService>();
+            }
 
-            KeyValueLogScope scope = Utils.CreateArtifactScope(Utils.ArtifactType_Dump, endpointInfo);
+            public async ValueTask DisposeAsync()
+            {
+                _disposalTokenSource.SafeCancel();
```

Comment:
Consider a _isDisposed flag that sets a disposed state and is used here as well as preventing calls to StartAync/WaitForCompletionAsync after disposal already happens.